### DISCUSSION TITLE
Beamer template improvements

### DISCUSSION THREAD

**Discussion by twsh** [jgm/pandoc#10237] [Ideas]
I have been thinking a bit about the Beamer template. I thought that I would mention some possible improvements. I could make PRs for any of these, if they seem like a good idea.

# Sections

The template includes this code:

```
\setbeamertemplate{part page}{
  \centering
  \begin{beamercolorbox}[sep=16pt,center]{part title}
    \usebeamerfont{part title}\insertpart\par
  \end{beamercolorbox}
}
\setbeamertemplate{section page}{
  \centering
  \begin{beamercolorbox}[sep=12pt,center]{section title}
    \usebeamerfont{section title}\insertsection\par
  \end{beamercolorbox}
}
\setbeamertemplate{subsection page}{
  \centering
  \begin{beamercolorbox}[sep=8pt,center]{subsection title}
    \usebeamerfont{subsection title}\insertsubsection\par
  \end{beamercolorbox}
}
```

The effect is to remove 'Part X' (and so on) from the part slides. I think that the reason for this was to avoid them being printed in English. I think that has been fixed by https://github.com/jgm/pandoc/issues/9868. So maybe this isn't needed any more?

# Theme options

The template has variables for theme, colortheme, fonttheme, innertheme, outertheme. There is also a variable for themeoptions. Why not for the rest?

# Short title/subtitle/author/institute/date

In Beamer, all of these take an optional argument for a short form. These are used by some themes, and many themes don't look good without them. So it might be good to have variables for them, unless this is bloating the template for little gain?

# Multiple titlegraphics

Perhaps the titlegraphic variable should take a list and produce a sequence of `\includegraphics`? Title graphic options should be applied to them all.

# Text in dates

The Beamer manual encourages using the `\date` command to insert the venue of a talk as well as the date. I don't know if this is just unnecessary complexity, but I thought about adding a 'venue' variable which is inserted into the `\date` command as well as the 'date' variable. Some logic could allow for either or both to be used. That keeps 'date' as a representation of the date which is better if the same source is also making, e.g., HTML. If this is a good idea, the other presentation formats could do the same thing.

# Professional fonts

Would it be a good idea to include `\usefonttheme{professionalfonts}` whenever the fontfamily variable is set? That's the impression I get from the Beamer manual, but I could be wrong.

# Font choice

The Beamer manual seems to gently discourage using Latin Modern (section 5.6.2). Is there a case for a different default, for example [arev.sty](https://ctan.org/pkg/arev)? There is the obvious downside that not everybody will like the change, and those that want it can use arev.sty anyway.

**Comment by samcarter** [comment]
> # Font choice
> 
> The Beamer manual seems to gently discourage using Latin Modern (section 5.6.2). Is there a case for a different default, for example [arev.sty](https://ctan.org/pkg/arev)? 

Most beamer users will just use the default computer modern sans font. This means that this is by far the most thoroughly tested font for beamer. If you'd like to minimise the risk of problems for your users, I would stick to the default font. 

(beamer also hard codes it in a couple of places, like operator names, which can lead to "funny" looking mixtures if combined with other fonts)

> There is the obvious downside that not everybody will like the change, and those that want it can use arev.sty anyway.

While it is very easy for a user to add their favourite font package to the header, it is hard to undo it and switch back to the beamer default behaviour.

**Comment by twsh** [comment]
That's good to know.

Out of interest, would that be a reason not to use lmodern.sty as well?

**Reply by jgm** [reply]
The beamer manual says:

> A small selection of alternatives to Computer Modern:
> • Latin Modern is a Computer Modern derivative that provides more characters, so it’s not considered a real alternative. It’s recommended over Computer Modern, though.

So I don't think they're saying "don't use Latin Modern."  They're saying: it's better than the default Computer Modern, but there are better fonts out there.

Seems to me that loading Latin Modern is a good plan.  You can still specify another mainfont or sansfont (and indeed some beamer themes, like metropolis, specify their own fonts).  But Latin Modern gives us a nice default that we can be pretty confident will be installed on the system.

**Comment by samcarter** [comment]
Looking at the code pandoc produces, there are a couple of things which could be improved

```
% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
```

unicode is the default now anyway

```
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
  ignorenonframetext,
]{beamer}
\usepackage{pgfpages}
```

pdfpages isn't necessary. Beamer will load it automatically if it is needed for notes

```
\setbeamertemplate{caption}[numbered]
\setbeamertemplate{caption label separator}{: }
\setbeamercolor{caption name}{fg=normal text.fg}
\beamertemplatenavigationsymbolsempty
% Prevent slide breaks in the middle of a paragraph
\widowpenalties 1 10000
\raggedbottom
\setbeamertemplate{part page}{
  \centering
  \begin{beamercolorbox}[sep=16pt,center]{part title}
    \usebeamerfont{part title}\insertpart\par
  \end{beamercolorbox}
}
\setbeamertemplate{section page}{
  \centering
  \begin{beamercolorbox}[sep=12pt,center]{section title}
    \usebeamerfont{section title}\insertsection\par
  \end{beamercolorbox}
}
\setbeamertemplate{subsection page}{
  \centering
  \begin{beamercolorbox}[sep=8pt,center]{subsection title}
    \usebeamerfont{subsection title}\insertsubsection\par
  \end{beamercolorbox}
}
\AtBeginPart{
  \frame{\partpage}
}
\AtBeginSection{
  \ifbibliography
  \else
    \frame{\sectionpage}
  \fi
}
\AtBeginSubsection{
  \frame{\subsectionpage}
}
\usepackage{amsmath,amssymb}
```
not necessary, beamer loads the ams* packages automatically

```
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
```

not necessary, utf8 is the default since 6 years or so

```
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\usetheme[]{Madrid}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
```

not necessary, loaded automatically


```
\newif\ifbibliography
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={My Title},
  pdfauthor={This Author},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{My Title}
\author{This Author}
\date{\today}

\begin{document}
\frame{\titlepage}

\section{Section 1}\label{section-1}

\begin{frame}{Slide 1}
\phantomsection\label{slide-1}
```

using `\begin{frame}[label=slide-1]\frametitle{Slide 1}` would allow your users more option to reference the label

```
Lorem ipsum
\end{frame}

\begin{frame}{Slide 2}
\phantomsection\label{slide-2}
Doler sit amen
\end{frame}

\end{document}
```

**Comment by jgm** [comment]
Commit 17b667ec260547201e422de7da56816d97d73c3b is where that first bit of code was added.
If I recall, the issue wasn't about the English language, but the numbering itself -- we shouldn't be numbering unless `--number-sections` is used.

**Reply by jgm** [reply]
I like some of these ideas, but hold off for a bit. I'm thinking of using partials to split the latex template into latex and beamer parts.  This should make it much more straightforward to modify the beamer template.

**Reply by jgm** [reply]
OK, I've pushed this commit, so now there is a separate default.beamer template.

This should make it easier to do beamer-specific customizations.

So, I'd say:  if anyone has specific ideas for improving the beamer template, feel free to make PRs, but each PR should propose one logical change, so they can be evaluated more easily one by one.

**Reply by twsh** [reply]
The new template organisation is much easier to follow. I will propose some of those changes soon.

**Reply by twsh** [reply]
I'm thinking now that there should be a `$if(numbersections)$` in both default.latex (not common.latex) and default.beamer, so that both formats support `--number-sections`. I had completely forgotten about that reason for the code, or maybe I didn't understand it properly when I reported the bug. It's on my list.

**Comment by twsh** [comment]
I made pull requests for the things that still seem to me to be good ideas.

I decided that the text in dates isn't very important, and that `$date-meta$` can be used to solve the issue.

Changing the font or using the professionalfonts theme also seem like more trouble than they are worth.
